### PR TITLE
feat(NPG refund): perform refund x minutes after auth request

### PR DESCRIPTION
**Pull request by pietro-tota**
Depends on https://github.com/pagopa/pagopa-ecommerce-local/pull/174

#### List of Changes
Add parameterized logic to posticipate refund event (visibility timeout) x minutes after authorization request have been requested for NPG transactions.
Since refund is performed writing an event on the queue the functionality uses the events visibility timeout to posticipate the refund request

#### Motivation and Context
NPG have an internal timeout that is used to close any pending authorization that have not been completed.
It is set to 5 minute, time after eCommerce will receive a POST notification with the authorization resulting outcome (CANCEL in case of pending authorization that have not been finalized).
Before this modification there are cases in which refund is processed before authorization process have been closed NPG side.
The case can be the following: an user start authorization process (that is, click on "Paga" button) close to the payment token expiration (actually set to 15 minute from transaction activation).
The transaction expiration is processed in background by the meaning of expiration event processing (15 minute after transaction activation) during authorization process (that can take several minutes if the payment is not frictionless).
This can lead to a situation where the user is finishing it's authorization and the refund mechanism is activated.
During refund processing for NPG authorizations a GET orders api call is made in order to retrieve authorization state before processing refund but for those cases the authorization is still pending and no action is performed, leading to transactions to be refund manually (user find transaction expired once redirected to checkout/io domains).

This case is covered taking the authorization requested event timestamp (written by eCommerce once authorization response from NPG is received) and adding a configurable offset (actually 10 minute, to be sure to be over NPG transaction authorization process): after this offset the authorization is completely processed and have reach a final state (that can be EXECUTED, CANCELED or any other failing status) and refund logic can then perform it's logic reasoning against a NPG authorization final status and perform automatic refund according to this state.

#### How Has This Been Tested?
Junit tests + local test with eCommerce local

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt [hunk 1]
```diff
@@ -406,30 +406,110 @@ fun requestRefundTransaction(
...
+      authorizationOutcome)
+    return Duration.ZERO
   }
+  val refundEventVisibilityTimeout =
+    if (paymentGateway == TransactionAuthorizationRequestData.PaymentGateway.NPG) {
+      val now = ZonedDateTime.now()
+      val processRefundNotBefore = authorizationRequestedTimestamp.plus(npgDelayFromAuthRequest)
+      val timeToWaitForRefund = Duration.between(now, processRefundNotBefore)
+      if (timeToWaitForRefund.isNegative) {
+        Duration.ZERO
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:499]:
```suggestion
      val processRefundNotBefore = authorizationRequestedTimestamp + npgDelayFromAuthRequest
      val timeToWaitForRefund = Duration.between(now, processRefundNotBefore)
      if (now.isAfter(processRefundNotBefore)) {
        Duration.ZERO
```

**Review by giovanniberti — APPROVED** [review]

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt [hunk 2]
```diff
@@ -406,30 +406,110 @@ fun requestRefundTransaction(
       IllegalArgumentException(
         "Tried to call `refundRequested` on transaction with null authorization request data in status ${transaction.status.value}!"))
   }
-
+  val authorizationRequestedTimestamp =
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt:409]: ops, code left here after refactoring 👀

[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt]: Fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/335/commits/facdbac039abe917b922495a77da3dac311280e2

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt (hunk 1 above)
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/common.kt]: Fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/335/commits/facdbac039abe917b922495a77da3dac311280e2

**Review by giovanniberti — APPROVED** [review]

**Pull request closed without merging**